### ISSUE TITLE
Children metadata is lost

### ISSUE THREAD

**Issue by mholo65** [statiqdev/Statiq.Framework#69]
I'm trying to build up Parent - Child relationship using `Children` key in some Code Analysis stuff in order to easier then build up navigation menu. But I'm facing problem where the relationship is lost. It may have to do with the special Metadata handling that documents outputted from in AnalyzeCSharp module has.

Anyway, here's a minimal repro:

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis;
using Microsoft.Extensions.Logging;
using Statiq.App;
using Statiq.CodeAnalysis;
using Statiq.Common;
using Statiq.Core;

namespace statiq_repro
{
    class Program
    {
        static Task Main(string[] args)
            =>  Bootstrapper
                .CreateDefault(args)
                .AddPipeline<BugPipeline>()
                .RunAsync();
    }

    internal class BugPipeline : Pipeline
    {
        public BugPipeline()
        {
            InputModules = new ModuleList
            {
                new GetFile()
            };

            ProcessModules = new ModuleList
            {
                new AnalyzeCSharp()
                        .WherePublic()
                        .WithAssemblySymbols()
                        .WithImplicitInheritDoc(),
                new CreateIndex(),
                new FindParent()
            };
        }
    }

    internal class FindParent : ParallelModule
    {
        protected override Task<IEnumerable<IDocument>> ExecuteInputAsync(IDocument input, IExecutionContext context)
        {
            var parent = input.GetParent(context.Inputs);

            context.LogInformation($"Parent for {input.ToSafeDisplayString()} is {parent.ToSafeDisplayString()}");

            return input.YieldAsync();
        }
    }

    internal class CreateIndex : ParallelModule
    {
        protected override Task<IEnumerable<IDocument>> ExecuteInputAsync(IDocument input, IExecutionContext context)
        {
            var symbolKind = input.Get<ISymbol>(CodeAnalysisKeys.Symbol).Kind;

            if (symbolKind != SymbolKind.Assembly &&
                symbolKind != SymbolKind.Namespace &&
                symbolKind != SymbolKind.NamedType)
            {
                return input.YieldAsync();
            }

            IReadOnlyList<IDocument> children;
            if (symbolKind == SymbolKind.Namespace)
            {
                // Children is all member types and member namespaces
                children = input.GetChildren(CodeAnalysisKeys.MemberTypes)
                    .Where(x => x.GetBool(CodeAnalysisKeys.IsResult) && x.Get<ISymbol>(CodeAnalysisKeys.Symbol).Kind == SymbolKind.NamedType)
                    .GroupBy(x => x.GetString(CodeAnalysisKeys.SpecificKind))
                    .OrderBy(x => x.Key)
                    .SelectMany(x => x.OrderBy(x => x.GetString(CodeAnalysisKeys.DisplayName)))
                    .Concat(input.GetChildren(CodeAnalysisKeys.MemberNamespaces)
                        .Where(x => x.GetBool(CodeAnalysisKeys.IsResult))
                        .OrderBy(x => x.GetString(CodeAnalysisKeys.DisplayName)))
                    .ToList()
                    .AsReadOnly();
            }
            else if (symbolKind == SymbolKind.Assembly)
            {
                // Children is all root namespaces
                children = context.Inputs
                    .Where(x => x.GetBool(CodeAnalysisKeys.IsResult) &&
                                x.Get<ISymbol>(CodeAnalysisKeys.Symbol).Kind == SymbolKind.Namespace &&
                                x.GetString(CodeAnalysisKeys.ContainingNamespace) is null)
                    .ToList()
                    .AsReadOnly();
            }
            else // Named types
            {
                // Children is all members
                children = input.GetChildren(CodeAnalysisKeys.Constructors)
                    .Where(x => x.GetBool(CodeAnalysisKeys.IsResult))
                    .OrderBy(x => x.GetString(CodeAnalysisKeys.DisplayName))
                    .Concat(input.GetChildren(CodeAnalysisKeys.Members)
                        .Where(x => x.GetBool(CodeAnalysisKeys.IsResult) && 
                                    (x.Get<ISymbol>(CodeAnalysisKeys.Symbol).Kind == SymbolKind.Event ||
                                     x.Get<ISymbol>(CodeAnalysisKeys.Symbol).Kind == SymbolKind.Field ||
                                     x.Get<ISymbol>(CodeAnalysisKeys.Symbol).Kind == SymbolKind.Property ||
                                     x.Get<ISymbol>(CodeAnalysisKeys.Symbol).Kind == SymbolKind.Method))
                        .OrderBy(x => x.GetString(CodeAnalysisKeys.Kind))
                        .ThenBy(x => x.GetString(CodeAnalysisKeys.DisplayName)))
                    .Concat(input.GetChildren(CodeAnalysisKeys.Operators)
                        .OrderBy(x => x.GetString(CodeAnalysisKeys.DisplayName)))
                    .Concat(input.GetChildren(CodeAnalysisKeys.ExtensionMethods)
                        .Where(x => x.GetBool(CodeAnalysisKeys.IsResult))
                        .OrderBy(x => x.GetString(CodeAnalysisKeys.DisplayName)))
                    .ToList()
                    .AsReadOnly();
            }

            context.LogInformation($"{input.ToSafeDisplayString()} has {children.Count} children, they are:\n{string.Join("\n", children.Select(x => x.ToSafeDisplayString()))}");

            return input.Clone(new MetadataDictionary(input)
            {
                [Keys.Children] = children
            }).YieldAsync();
        }
    }

    internal class GetFile : IModule
    {
        public async Task<IEnumerable<IDocument>> ExecuteAsync(IExecutionContext context)
        {
            const string content = @"
                namespace Foo
                {
                    public class Blue
                    {
                        public void Green()
                        {
                        }

                        public void Red()
                        {
                        }
                    }
                }";

            return context.CreateDocument(await context.GetContentProviderAsync(content).ConfigureAwait(false)).Yield();
        }
    }
}
```

The output from the above code is:
```
[INFO] Statiq version 1.0.0-alpha.19
[INFO] Root path:
       C:/src/tmp/statiq-repro
[INFO] Input path(s):
       theme
       input
[INFO] Output path:
       output
[INFO] Temp path:
       temp
[INFO] Executing 1 pipelines (BugPipeline)
[INFO] Cleaned temp directory: temp
[INFO] Cleaned output directory: output
[INFO] -> BugPipeline/Input » Starting BugPipeline Input phase execution... (0 input document(s), 1 module(s))
[INFO]    BugPipeline/Input » Finished BugPipeline Input phase execution (1 output document(s), 27 ms)
[INFO] -> BugPipeline/Process » Starting BugPipeline Process phase execution... (1 input document(s), 3 module(s))
[INFO] BugPipeline/Process » CreateIndex » <global namespace> has 1 children, they are:
Foo
[INFO] BugPipeline/Process » CreateIndex » Foo has 1 children, they are:
Foo.Blue
[INFO] BugPipeline/Process » CreateIndex » Foo.Blue has 2 children, they are:
Foo.Blue.Green()
Foo.Blue.Red()
[INFO] BugPipeline/Process » FindParent » Parent for Foo.Blue.Red() is Foo.Blue
[INFO] BugPipeline/Process » FindParent » Parent for Foo.Blue.Green() is Foo.Blue
[INFO] BugPipeline/Process » FindParent » Parent for Foo.Blue is null
[INFO] BugPipeline/Process » FindParent » Parent for <global namespace> is null
[INFO] BugPipeline/Process » FindParent » Parent for Foo is null
[INFO]    BugPipeline/Process » Finished BugPipeline Process phase execution (5 output document(s), 1038 ms)
[INFO] Execution summary: (number of output documents per pipeline and phase)

 | Pipeline    | Input     | Process     | Transform | Output | Total Time |
 |-------------------------------------------------------------------------|
 | BugPipeline | 1 (27 ms) | 5 (1038 ms) |           |        | 1065 ms    |

[INFO] Finished execution in 1098 ms
[INFO] Cleaned temp directory: temp
```

As you can see, when creating the relationship, the children are found. But when then trying to search for parent in the next module, I can only find parent for the methods.

**Comment by mholo65** [comment]
To fix this, I have created a `DocumentComparer`, which compares by `Id`, e.g. change the `FindParent` module to:

```csharp
    internal class FindParent : ParallelModule
    {
        private class DocumentComparer : IEqualityComparer<IDocument>
        {
            private static readonly Lazy<DocumentComparer> _instance = new Lazy<DocumentComparer>(() => new DocumentComparer());

            public static DocumentComparer Instance => _instance.Value;

            public bool Equals(IDocument x, IDocument y) => x?.Id == y?.Id;

            public int GetHashCode(IDocument obj) => obj?.Id.GetHashCode() ?? 0;
        }

        protected override Task<IEnumerable<IDocument>> ExecuteInputAsync(IDocument input, IExecutionContext context)
        {
            var parent = GetParent(input, context.Inputs);

            context.LogInformation($"Parent for {input.ToSafeDisplayString()} is {parent.ToSafeDisplayString()}");

            return input.YieldAsync();
        }

        private IDocument GetParent(IDocument document, IEnumerable<IDocument> parents, bool recursive = true, string key = Keys.Children)
        {
            _ = parents ?? throw new ArgumentNullException(nameof(parents));
            _ = key ?? throw new ArgumentNullException(nameof(key));

            IDocument parent = parents.FirstOrDefault(x => x.GetChildren(key).Contains(document, DocumentComparer.Instance));
            if (parent == null && recursive)
            {
                foreach (IDocument candidate in parents)
                {
                    parent = GetParent(document, candidate.GetChildren(key), true, key);
                    if (parent != null)
                    {
                        break;
                    }
                }
            }
            return parent;
        }
    }
```

Now the output is

```csharp
[INFO] Statiq version 1.0.0-alpha.20
[INFO] Root path:
       C:/src/tmp/statiq-repro
[INFO] Input path(s):
       theme
       input
[INFO] Output path:
       output
[INFO] Temp path:
       temp
[INFO] Executing 1 pipelines (BugPipeline)
[INFO] Cleaned temp directory: temp
[INFO] Cleaned output directory: output
[INFO] -> BugPipeline/Input » Starting BugPipeline Input phase execution... (0 input document(s), 1 module(s))
[INFO]    BugPipeline/Input » Finished BugPipeline Input phase execution (1 output document(s), 25 ms)
[INFO] -> BugPipeline/Process » Starting BugPipeline Process phase execution... (1 input document(s), 3 module(s))
[INFO] BugPipeline/Process » CreateIndex » Foo has 1 children, they are:
Foo.Blue
[INFO] BugPipeline/Process » CreateIndex » <global namespace> has 1 children, they are:
Foo
[INFO] BugPipeline/Process » CreateIndex » Foo.Blue has 2 children, they are:
Foo.Blue.Green()
Foo.Blue.Red()
[INFO] BugPipeline/Process » FindParent » Parent for Foo is <global namespace>
[INFO] BugPipeline/Process » FindParent » Parent for Foo.Blue is Foo
[INFO] BugPipeline/Process » FindParent » Parent for Foo.Blue.Green() is Foo.Blue
[INFO] BugPipeline/Process » FindParent » Parent for Foo.Blue.Red() is Foo.Blue
[INFO] BugPipeline/Process » FindParent » Parent for <global namespace> is null
[INFO]    BugPipeline/Process » Finished BugPipeline Process phase execution (5 output document(s), 1031 ms)
[INFO] Execution summary: (number of output documents per pipeline and phase)

 | Pipeline    | Input     | Process     | Transform | Output | Total Time |
 |-------------------------------------------------------------------------|
 | BugPipeline | 1 (25 ms) | 5 (1031 ms) |           |        | 1056 ms    |

[INFO] Finished execution in 1094 ms
[INFO] Cleaned temp directory: temp
```

**Comment by daveaglick** [comment]
I think the problem here is one of immutability and the way it impacts relationship-based structures. Since documents are immutable, changing anything about them (like metadata) creates a new instance. Any metadata that document had that points to other documents is still intact, but the opposite isn't true. If other documents had a reference to the document that was just changed (and therefore cloned) those references don't update to the new clone and they're now pointing to the old document.

This is where the `Id` property comes in. A document keeps the same ID from one clone to the next. It lets you "rebuild" those relationships later by comparing the IDs of the originally referenced instance with the set of newly cloned instances to find the most recent "version" of that document. That appears to be exactly what you're doing with the help of the comparer.

TL;DR: would adding the `DocumentComparer` to `Statiq.Core` be helpful? You'd still have to do the manual matching of the original children with the newly cloned ones, but at least the comparer would be provided. I'm hesitant to make default document equality use ID because I suspect instance equality is still pretty important for documents (but could be convinced otherwise).

**Comment by daveaglick** [comment]
I went ahead and added a `DocumentIdComparer` to `Statiq.Common`, otherwise I think this behavior is as designed.
